tests/sphere_test.cpp

```diff
@@ -0,0 +1,104 @@
+  Ray alongYAxisOppposite(behindSphereYAxis, Vec3::kY);
+  ASSERT_FALSE(sphere.hitObject(alongYAxisOppposite, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+
+  Vec3 arbitraryVec3(12.123, 5.623, 13.235);
+  ASSERT_FALSE(sphere.isVecInside(arbitraryVec3));
+  Ray arbitraryOriginRay(arbitraryVec3, sphere.getOrigin() - arbitraryVec3);
+  ASSERT_TRUE(sphere.hitObject(arbitraryOriginRay, kSkipFreq, &hitData));
+  ASSERT_NEAR((arbitraryVec3 - sphere.getOrigin()).magnitude() -
+                  sphere.getRadius(),
+              hitData.time, constants::kAccuracy);
+
+  Vec3 alongXOneHitOrigin = Vec3(-5, 0, 1);
+  ASSERT_FALSE(sphere.isVecInside(alongXOneHitOrigin));
+  Ray alongXOneHit(alongXOneHitOrigin, Vec3::kX);
+  ASSERT_TRUE(sphere.hitObject(alongXOneHit, kSkipFreq, &hitData));
+  ASSERT_EQ(5, hitData.time);
+  ASSERT_EQ(Vec3(0, 0, 1), hitData.collisionPoint());
+
+  Vec3 alongXOneHitOriginMiss = Vec3(-5, 0, 1 + constants::kAccuracy);
+  ASSERT_FALSE(sphere.isVecInside(alongXOneHitOriginMiss));
+  Ray alongXOneHitMiss(alongXOneHitOriginMiss, Vec3::kX);
+  ASSERT_FALSE(sphere.hitObject(alongXOneHitMiss, kSkipFreq, &hitData));
+  ASSERT_EQ(5, hitData.time);
+  ASSERT_EQ(Vec3(0, 0, 1), hitData.collisionPoint());
+}
+
+TEST(SphereCollisionTest, RayHitInsideSphere) {
+  Sphere sphere(Vec3::kZero, 1);
+  RayHitData hitData;
+
+  Ray alongXAxis(Vec3::kZero, Vec3::kX);
+  ASSERT_TRUE(sphere.hitObject(alongXAxis, kSkipFreq, &hitData));
+  ASSERT_FLOAT_EQ(sphere.getRadius(), hitData.time);
+
+  Vec3 arbitraryChosenOriginInsideSphere(0.213, 0.523, 0.123);
+  ASSERT_TRUE(sphere.isVecInside(arbitraryChosenOriginInsideSphere));
+  Ray arbitraryRayInsideSphere(arbitraryChosenOriginInsideSphere,
+                               Vec3(214.2345, 2345, 2235.456));
+  ASSERT_TRUE(sphere.hitObject(arbitraryRayInsideSphere, kSkipFreq, &hitData));
+
+  Vec3 arbitraryInsideSphere = Vec3(12, 0.345, 3.457).normalize() * 0.99;
+  ASSERT_TRUE(sphere.isVecInside(arbitraryInsideSphere));
+  Ray randomInside(arbitraryInsideSphere, Vec3::kY);
+  ASSERT_TRUE(sphere.hitObject(randomInside, kSkipFreq, &hitData));
+}
+
+TEST(SphereCollisionTest, RayAtEdgeOfSphere) {
+  Sphere sphere(Vec3::kZero, 1);
+  RayHitData hitData;
+
+  Ray alongXAxis(Vec3::kX, Vec3::kX);
+  ASSERT_FALSE(sphere.hitObject(alongXAxis, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+  Ray alongXAxisHit(Vec3(1 - constants::kAccuracy, 0, 0), Vec3::kX);
+  ASSERT_TRUE(sphere.hitObject(alongXAxisHit, kSkipFreq, &hitData));
+  ASSERT_NEAR(constants::kAccuracy, hitData.time, constants::kAccuracy / 2);
+
+  // Prevents of getting ray inside to the sphere by
+  // floating point number error
+  Ray toOriginOfSphere(Vec3::kX, -Vec3::kX);
+  ASSERT_FALSE(sphere.hitObject(toOriginOfSphere, kSkipFreq, &hitData));
+  Vec3 xOffset(constants::kAccuracy, 0, 0);
+  Ray toOriginOfSphereFromOutside(Vec3::kX + 1.1 * xOffset, -Vec3::kX);
```

Comment:
Where is 1.1 coming from? Why not 20?

x1.1 does not seem to be on the edge any more.

Reply:
After all considerations I dont think these test are usefull actually. I tried to do here a test, when offset is bigger then constants::kAccuracy by a little bit, it hits the sphere. Theoretically simple 1.0 * xOffset should work, but because of floafting point error it doesn't, so Im deleting it